In [1]:
# Import required libraries
import pandas as pd
import numpy as np  
import matplotlib.pyplot as plt
import seaborn as sns


# Libraries for data preprocessing and modeling
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer

# Machine Learning Models
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier

# Evaluation Metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

In [2]:
from pathlib import Path

data_path = Path('../data/diabetes_prediction_dataset.csv')
if not data_path.exists():
    data_path = Path('data/diabetes_prediction_dataset.csv')

df = pd.read_csv(data_path)
df.head()

,gender,age,hypertension,heart_disease,smoking_history,bmi,HbA1c_level,blood_glucose_level,diabetes
0,Female,80.0,0,1,never,25.19,6.6,140,0
1,Female,54.0,0,0,No Info,27.32,6.6,80,0
2,Male,28.0,0,0,never,27.32,5.7,158,0
3,Female,36.0,0,0,current,23.45,5.0,155,0
4,Male,76.0,1,1,current,20.14,4.8,155,0


Preparing X and Y variable

In [3]:
# preprare the data for modeling
X = df.drop('diabetes', axis=1)  # Features
y = df['diabetes']  # Target variable

print("Features (X) shape:", X.shape)
print("Target (y) shape:", y.shape)

Features (X) shape: (100000, 8)
Target (y) shape: (100000,)


Prepocessing the data

In [4]:
# Prepocessing the data
# Identify numeric and categorical columns

numeric_features = X.select_dtypes(exclude=['object']).columns
categorical_features = X.select_dtypes(include=['object']).columns 

print("Numeric features:", numeric_features.tolist())
print("Categorical features:", categorical_features.tolist())

Numeric features: ['age', 'hypertension', 'heart_disease', 'bmi', 'HbA1c_level', 'blood_glucose_level']
Categorical features: ['gender', 'smoking_history']


In [5]:
# Define transformers for numeric and categorical features
# For numeric features, we will use StandardScaler to standardize the data.
# For categorical features, we will use OneHotEncoder to convert them into a format suitable for machine learning algorithms.
numeric_transformer = StandardScaler()
categorical_transformer = OneHotEncoder(handle_unknown='ignore') 

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numeric_features),
        ('cat', categorical_transformer, categorical_features)
    ]
)

In [6]:
X = preprocessor.fit_transform(X)
print("Preprocessed features shape:", X.shape)

Preprocessed features shape: (100000, 15)


In [7]:
# Separate the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("Training features shape:", X_train.shape)
print("Testing features shape:", X_test.shape)
print("-"*30)
print("Training labels shape:", y_train.shape)
print("Testing labels shape:", y_test.shape)

Training features shape: (80000, 15)
Testing features shape: (20000, 15)
------------------------------
Training labels shape: (80000,)
Testing labels shape: (20000,)


Create an Evaluate Function to give all metrics after model Training

In [8]:
# Define the model evaluation function


def evaluate_model(true,predicted):
    ac_score = accuracy_score(true, predicted)
    pr_score = precision_score(true, predicted)
    re_score = recall_score(true, predicted)
    f1 = f1_score(true, predicted)
    

    return ac_score, pr_score, re_score, f1 

In [9]:
# Model
models= {
    'Logistic Regression': LogisticRegression(random_state=42),
    'Random Forest': RandomForestClassifier(random_state=42),
    'Decision Tree': DecisionTreeClassifier(random_state=42),
    'Gradient Boosting': GradientBoostingClassifier(random_state=42),
    'K-Nearest Neighbors': KNeighborsClassifier()
}

In [10]:
# Train, predict and evaluate each model

for name, model in models.items():
    
    model.fit(X_train, y_train)

    # Make predictions
    y_pred = model.predict(X_test)
    y_test_pred = model.predict(X_train)

    # Evaluate Train and Test dataset

    train_ac, train_pr, train_re, train_f1 = evaluate_model(y_train, y_test_pred)
    test_ac, test_pr, test_re, test_f1 = evaluate_model(y_test, y_pred)
    

    # Print the evaluation results
    print(f"Model: {name.upper()}")
    print("Model Performance on Train Dataset:")
    print(f"Accuracy: {train_ac:.4f}")
    print(f"Precision: {train_pr:.4f}")
    print(f"Recall: {train_re:.4f}")
    print(f"F1 Score: {train_f1:.4f}")

    print("-"*30)

    print("Model Performance on Test Dataset:")
    print(f"Accuracy: {test_ac:.4f}")
    print(f"Precision: {test_pr:.4f}")
    print(f"Recall: {test_re:.4f}")
    print(f"F1 Score: {test_f1:.4f}")

    print("="*50, "\n")

Model: LOGISTIC REGRESSION
Model Performance on Train Dataset:
Accuracy: 0.9607
Precision: 0.8692
Recall: 0.6321
F1 Score: 0.7319
------------------------------
Model Performance on Test Dataset:
Accuracy: 0.9589
Precision: 0.8631
Recall: 0.6165
F1 Score: 0.7193

Model: RANDOM FOREST
Model Performance on Train Dataset:
Accuracy: 0.9993
Precision: 0.9993
Recall: 0.9923
F1 Score: 0.9958
------------------------------
Model Performance on Test Dataset:
Accuracy: 0.9700
Precision: 0.9468
Recall: 0.6879
F1 Score: 0.7969

Model: DECISION TREE
Model Performance on Train Dataset:
Accuracy: 0.9993
Precision: 1.0000
Recall: 0.9919
F1 Score: 0.9959
------------------------------
Model Performance on Test Dataset:
Accuracy: 0.9529
Precision: 0.7231
Recall: 0.7278
F1 Score: 0.7254

Model: GRADIENT BOOSTING
Model Performance on Train Dataset:
Accuracy: 0.9723
Precision: 0.9876
Recall: 0.6821
F1 Score: 0.8069
------------------------------
Model Performance on Test Dataset:
Accuracy: 0.9724
Precision

#### Model Performance Comparison (Test Dataset)

| Model | Accuracy | Precision | Recall | F1 Score |
|------|------|------|------|------|
| Logistic Regression | 0.9589 | 0.8631 | 0.6165 | 0.7193 |
| Random Forest | 0.9699 | 0.9417 | 0.6903 | 0.7966 |
| Decision Tree | 0.9532 | 0.7206 | **0.7383** | 0.7293 |
| Gradient Boosting | **0.9724** | **0.9874** | 0.6856 | **0.8093** |
| K-Nearest Neighbors | 0.9607 | 0.8948 | 0.6124 | 0.7271 |



- Among these models, we can see Gradient Boosting model achieved  the best overall performance with the highest accuracy (97.24%) and F1 score (0.8093) on the test dataset. 

- Therefore, Gradient Boosting is selected as the final model for diabetes prediction

